File: benchmarks/src/validation.ts
/**
 * Validation module for benchmark correctness
 * Validates that numpy-ts produces the same results as NumPy
 */

import { spawn } from 'child_process';
import { resolve } from 'path';
import * as np from '../../src/index';
import type { BenchmarkCase } from './types';

const FLOAT_TOLERANCE = 1e-10;

/**
 * Compare two arrays or scalars for equality with tolerance
 */
function resultsMatch(numpytsResult: any, numpyResult: any): boolean {
  // Both scalars
  if (typeof numpytsResult === 'number' && typeof numpyResult === 'number') {
    return Math.abs(numpytsResult - numpyResult) < FLOAT_TOLERANCE;
  }

  if (typeof numpytsResult === 'boolean' && typeof numpyResult === 'boolean') {
    return numpytsResult === numpyResult;
  }

  // Both arrays (both should be {shape, data} format at this point)
  if (numpytsResult?.shape && numpyResult?.shape) {
    // Check shapes match
    if (numpytsResult.shape.length !== numpyResult.shape.length) {
      return false;
    }
    if (!numpytsResult.shape.every((dim: number, i: number) => dim === numpyResult.shape[i])) {
      return false;
    }

    // Compare values element by element
    // Both are already in {shape, data} format
    const tsData = numpytsResult.data;
    const npData = numpyResult.data;

    return arraysEqual(tsData, npData);
  }

  return false;
}

/**
 * Recursively compare nested arrays with tolerance
 */
function arraysEqual(a: any, b: any): boolean {
  if (Array.isArray(a) && Array.isArray(b)) {
    if (a.length !== b.length) return false;
    return a.every((val, i) => arraysEqual(val, b[i]));
  }

  // Compare numbers with tolerance
  if (typeof a === 'number' && typeof b === 'number') {
    if (isNaN(a) && isNaN(b)) return true;
    if (!isFinite(a) && !isFinite(b)) return a === b; // Both inf or -inf
    return Math.abs(a - b) < FLOAT_TOLERANCE;
  }

  // Compare booleans
  if (typeof a === 'boolean' && typeof b === 'boolean') {
    return a === b;
  }

  return a === b;
}

/**
 * Run a single benchmark operation with numpy-ts
 */
function runNumpyTsOperation(spec: BenchmarkCase): any {
  // Setup arrays
  const arrays: Record<string, any> = {};

  for (const [key, config] of Object.entries(spec.setup)) {
    const { shape, dtype = 'float64', fill = 'zeros', value } = config;

    // Handle scalar values
    if (['n', 'axis', 'new_shape', 'shape', 'fill_value', 'target_shape'].includes(key)) {
      arrays[key] = shape[0];
      if (key === 'new_shape' || key === 'shape' || key === 'target_shape') {
        arrays[key] = shape;
      }
      continue;
    }

    // Handle indices array
    if (key === 'indices') {
      arrays[key] = shape;
      continue;
    }

    // Create arrays
    if (value !== undefined) {
      arrays[key] = np.full(shape, value, dtype);
    } else if (fill === 'zeros') {
      arrays[key] = np.zeros(shape, dtype);
    } else if (fill === 'ones') {
      arrays[key] = np.ones(shape, dtype);
    } else if (fill === 'random' || fill === 'arange') {
      const size = shape.reduce((a, b) => a * b, 1);
      const flat = np.arange(0, size, 1, dtype);
      arrays[key] = flat.reshape(...shape);
    }
  }

  // Execute operation
  switch (spec.operation) {
    // Creation
    case 'zeros':
      return np.zeros(arrays.shape);
    case 'ones':
      return np.ones(arrays.shape);
    case 'arange':
      return np.arange(0, arrays.n);
    case 'linspace':
      return np.linspace(0, 100, arrays.n);
    case 'logspace':
      return np.logspace(0, 3, arrays.n);
    case 'geomspace':
      return np.geomspace(1, 1000, arrays.n);
    case 'eye':
      return np.eye(arrays.n);
    case 'identity':
      return np.identity(arrays.n);
    case 'empty':
      return np.empty(arrays.shape);
    case 'full':
      return np.full(arrays.shape, arrays.fill_value);
    case 'copy':
      return arrays.a.copy();
    case 'zeros_like':
      return np.zeros_like(arrays.a);

    // Arithmetic
    case 'add':
      return arrays.b !== undefined ? arrays.a.add(arrays.b) : arrays.a.add(arrays.scalar);
    case 'multiply':
      return arrays.b !== undefined
        ? arrays.a.multiply(arrays.b)
        : arrays.a.multiply(arrays.scalar);
    case 'mod':
      return np.mod(arrays.a, arrays.b || arrays.scalar);
    case 'floor_divide':
      return np.floor_divide(arrays.a, arrays.b || arrays.scalar);
    case 'reciprocal':
      return np.reciprocal(arrays.a);

    // Math
    case 'sqrt':
      return np.sqrt(arrays.a);
    case 'power':
      return np.power(arrays.a, 2);
    case 'absolute':
      return np.absolute(arrays.a);
    case 'negative':
      return np.negative(arrays.a);
    case 'sign':
      return np.sign(arrays.a);

    // Linalg
    case 'dot':
      return arrays.a.dot(arrays.b);
    case 'inner':
      return arrays.a.inner(arrays.b);
    case 'outer':
      return arrays.a.outer(arrays.b);
    case 'tensordot':
      return arrays.a.tensordot(arrays.b, arrays.axes ?? 2);
    case 'matmul':
      return arrays.a.matmul(arrays.b);
    case 'trace':
      return arrays.a.trace();
    case 'transpose':
      return arrays.a.transpose();

    // Reductions
    case 'sum':
      return arrays.axis !== undefined ? arrays.a.sum(arrays.axis) : arrays.a.sum();
    case 'mean':
      return arrays.a.mean();
    case 'max':
      return arrays.a.max();
    case 'min':
      return arrays.a.min();
    case 'prod':
      return arrays.a.prod();
    case 'argmin':
      return arrays.a.argmin();
    case 'argmax':
      return arrays.a.argmax();
    case 'var':
      return arrays.a.var();
    case 'std':
      return arrays.a.std();
    case 'all':
      return arrays.a.all();
    case 'any':
      return arrays.a.any();

    // Reshape
    case 'reshape':
      return arrays.a.reshape(...arrays.new_shape);
    case 'flatten':
      return arrays.a.flatten();
    case 'ravel':
      return arrays.a.ravel();

    // Array manipulation
    case 'swapaxes':
      return np.swapaxes(arrays.a, 0, 1);
    case 'concatenate':
      return np.concatenate([arrays.a, arrays.b], 0);
    case 'stack':
      return np.stack([arrays.a, arrays.b], 0);
    case 'vstack':
      return np.vstack([arrays.a, arrays.b]);
    case 'hstack':
      return np.hstack([arrays.a, arrays.b]);
    case 'tile':
      return np.tile(arrays.a, [2, 2]);
    case 'repeat':
      return arrays.a.repeat(2);

    // Advanced
    case 'broadcast_to':
      return np.broadcast_to(arrays.a, arrays.target_shape);
    case 'take':
      return arrays.a.take(arrays.indices);

    default:
      throw new Error(`Unknown operation: ${spec.operation}`);
  }
}

/**
 * Validate all benchmarks produce correct results
 */
export async function validateBenchmarks(specs: BenchmarkCase[]): Promise<void> {
  const scriptPath = resolve(__dirname, '../scripts/validation.py');

  return new Promise((resolve, reject) => {
    const python = spawn('python3', [scriptPath]);

    let stdout = '';
    let stderr = '';

    python.stdout.on('data', (data) => {
      stdout += data.toString();
    });

    python.stderr.on('data', (data) => {
      stderr += data.toString();
    });

    python.on('close', (code) => {
      if (code !== 0) {
        reject(new Error(`Validation script failed: ${stderr}`));
        return;
      }

      try {
        const numpyResults = JSON.parse(stdout);

        // Run numpy-ts operations and compare
        let passed = 0;
        let failed = 0;

        for (let i = 0; i < specs.length; i++) {
          const spec = specs[i]!;
          const numpyResult = numpyResults[i];

          try {
            const numpytsResult = runNumpyTsOperation(spec);

            // Convert numpy-ts result to comparable format
            let tsValue: any;
            if (typeof numpytsResult === 'number' || typeof numpytsResult === 'boolean') {
              tsValue = numpytsResult;
            } else if (
              numpytsResult &&
              typeof numpytsResult === 'object' &&
              'shape' in numpytsResult
            ) {
              // It's an NDArray - check if it has toArray method
              if (typeof numpytsResult.toArray !== 'function') {
                throw new Error(
                  `NDArray missing toArray method. Type: ${typeof numpytsResult}, keys: ${Object.keys(numpytsResult).join(', ')}`
                );
              }
              tsValue = {
                shape: Array.from(numpytsResult.shape),
                data: numpytsResult.toArray(),
              };
            } else {
              tsValue = numpytsResult;
            }

            if (resultsMatch(tsValue, numpyResult)) {
              passed++;
            } else {
              failed++;
              console.error(`  ❌ ${spec.name}: Results don't match`);
              console.error(`     numpy-ts: ${JSON.stringify(tsValue).substring(0, 100)}`);
              console.error(`     NumPy:    ${JSON.stringify(numpyResult).substring(0, 100)}`);
            }
          } catch (err) {
            failed++;
            console.error(`  ❌ ${spec.name}: Error - ${err}`);
          }
        }

        if (failed > 0) {
          reject(
            new Error(
              `Validation failed: ${failed}/${specs.length} benchmarks produced incorrect results`
            )
          );
        } else {
          console.log(`  ✓ ${passed}/${specs.length} operations validated`);
          resolve();
        }
      } catch (err) {
        reject(new Error(`Failed to parse validation output: ${err}`));
      }
    });

    python.on('error', (err) => {
      reject(new Error(`Failed to spawn Python: ${err.message}`));
    });

    // Send specs to Python
    python.stdin.write(JSON.stringify({ specs }));
    python.stdin.end();
  });
}

File: benchmarks/scripts/validation.py
#!/usr/bin/env python3
"""
Validation script for benchmark correctness
Runs operations with NumPy and returns results for comparison
"""

import json
import sys
import numpy as np


def setup_arrays(setup_config):
    """Create NumPy arrays from setup configuration"""
    arrays = {}

    for key, config in setup_config.items():
        shape = config["shape"]
        dtype = config.get("dtype", "float64")
        fill = config.get("fill", "zeros")
        value = config.get("value")

        # Handle scalar values
        if key in ["n", "axis", "new_shape", "shape", "fill_value", "target_shape"]:
            arrays[key] = shape[0]
            if key in ["new_shape", "shape", "target_shape"]:
                arrays[key] = shape
            continue

        # Handle indices array
        if key == "indices":
            arrays[key] = shape
            continue

        # Map dtype names
        np_dtype = dtype
        if dtype == "bool":
            np_dtype = "bool_"

        # Create arrays
        if value is not None:
            arrays[key] = np.full(shape, value, dtype=np_dtype)
        elif fill == "zeros":
            arrays[key] = np.zeros(shape, dtype=np_dtype)
        elif fill == "ones":
            arrays[key] = np.ones(shape, dtype=np_dtype)
        elif fill in ["random", "arange"]:
            size = np.prod(shape)
            flat = np.arange(0, size, 1, dtype=np_dtype)
            arrays[key] = flat.reshape(shape)

    return arrays


def run_operation(spec):
    """Run a single operation and return result"""
    arrays = setup_arrays(spec["setup"])
    operation = spec["operation"]

    # Execute operation
    if operation == "zeros":
        result = np.zeros(arrays["shape"])
    elif operation == "ones":
        result = np.ones(arrays["shape"])
    elif operation == "arange":
        result = np.arange(0, arrays["n"])
    elif operation == "linspace":
        result = np.linspace(0, 100, arrays["n"])
    elif operation == "logspace":
        result = np.logspace(0, 3, arrays["n"])
    elif operation == "geomspace":
        result = np.geomspace(1, 1000, arrays["n"])
    elif operation == "eye":
        result = np.eye(arrays["n"])
    elif operation == "identity":
        result = np.identity(arrays["n"])
    elif operation == "empty":
        result = np.empty(arrays["shape"])
        # For empty, just return zeros (we can't compare uninitialized data)
        result = np.zeros(arrays["shape"])
    elif operation == "full":
        result = np.full(arrays["shape"], arrays["fill_value"])
    elif operation == "copy":
        result = arrays["a"].copy()
    elif operation == "zeros_like":
        result = np.zeros_like(arrays["a"])

    # Arithmetic
    elif operation == "add":
        result = arrays["a"] + (
            arrays.get("b") if "b" in arrays else arrays.get("scalar")
        )
    elif operation == "multiply":
        result = arrays["a"] * (
            arrays.get("b") if "b" in arrays else arrays.get("scalar")
        )
    elif operation == "mod":
        divisor = arrays.get("b") if "b" in arrays else arrays.get("scalar")
        result = np.mod(arrays["a"], divisor)
    elif operation == "floor_divide":
        divisor = arrays.get("b") if "b" in arrays else arrays.get("scalar")
        result = np.floor_divide(arrays["a"], divisor)
    elif operation == "reciprocal":
        result = np.reciprocal(arrays["a"])

    # Math
    elif operation == "sqrt":
        result = np.sqrt(arrays["a"])
    elif operation == "power":
        result = np.power(arrays["a"], 2)
    elif operation == "absolute":
        result = np.absolute(arrays["a"])
    elif operation == "negative":
        result = np.negative(arrays["a"])
    elif operation == "sign":
        result = np.sign(arrays["a"])

    # Trigonometric
    elif operation == "sin":
        result = np.sin(arrays["a"])
    elif operation == "cos":
        result = np.cos(arrays["a"])
    elif operation == "tan":
        result = np.tan(arrays["a"])
    elif operation == "arctan2":
        result = np.arctan2(arrays["a"], arrays["b"])
    elif operation == "hypot":
        result = np.hypot(arrays["a"], arrays["b"])

    # Hyperbolic
    elif operation == "sinh":
        result = np.sinh(arrays["a"])
    elif operation == "cosh":
        result = np.cosh(arrays["a"])
    elif operation == "tanh":
        result = np.tanh(arrays["a"])

    # Linalg
    elif operation == "dot":
        result = np.dot(arrays["a"], arrays["b"])
    elif operation == "inner":
        result = np.inner(arrays["a"], arrays["b"])
    elif operation == "outer":
        result = np.outer(arrays["a"], arrays["b"])
    elif operation == "matmul":
        result = arrays["a"] @ arrays["b"]
    elif operation == "trace":
        result = np.trace(arrays["a"])
    elif operation == "transpose":
        result = arrays["a"].T

    # Reductions
    elif operation == "sum":
        result = arrays["a"].sum(axis=arrays.get("axis"))
    elif operation == "mean":
        result = arrays["a"].mean()
    elif operation == "max":
        result = arrays["a"].max()
    elif operation == "min":
        result = arrays["a"].min()
    elif operation == "prod":
        result = arrays["a"].prod()
    elif operation == "argmin":
        result = arrays["a"].argmin()
    elif operation == "argmax":
        result = arrays["a"].argmax()
    elif operation == "var":
        result = arrays["a"].var()
    elif operation == "std":
        result = arrays["a"].std()
    elif operation == "all":
        result = arrays["a"].all()
    elif operation == "any":
        result = arrays["a"].any()

    # Reshape
    elif operation == "reshape":
        result = arrays["a"].reshape(*arrays["new_shape"])
    elif operation == "flatten":
        result = arrays["a"].flatten()
    elif operation == "ravel":
        result = arrays["a"].ravel()

    # Array manipulation
    elif operation == "swapaxes":
        result = np.swapaxes(arrays["a"], 0, 1)
    elif operation == "concatenate":
        result = np.concatenate([arrays["a"], arrays["b"]], axis=0)
    elif operation == "stack":
        result = np.stack([arrays["a"], arrays["b"]], axis=0)
    elif operation == "vstack":
        result = np.vstack([arrays["a"], arrays["b"]])
    elif operation == "hstack":
        result = np.hstack([arrays["a"], arrays["b"]])
    elif operation == "tile":
        result = np.tile(arrays["a"], [2, 2])
    elif operation == "repeat":
        result = np.repeat(arrays["a"], 2)

    # Advanced
    elif operation == "broadcast_to":
        result = np.broadcast_to(arrays["a"], arrays["target_shape"])
    elif operation == "take":
        result = np.take(arrays["a"], arrays["indices"])

    else:
        raise ValueError(f"Unknown operation: {operation}")

    # Convert result to JSON-serializable format
    if isinstance(result, np.ndarray):
        return {"shape": result.shape, "data": result.tolist()}
    elif isinstance(result, (np.integer, np.floating)):
        return float(result)
    elif isinstance(result, np.bool_):
        return bool(result)
    else:
        return result


import math

def serialize_value(val):
    """Recursively serialize values, handling Infinity and NaN"""
    if isinstance(val, dict):
        return {k: serialize_value(v) for k, v in val.items()}
    elif isinstance(val, (list, tuple)):
        return [serialize_value(v) for v in val]
    elif isinstance(val, float):
        if math.isnan(val):
            return "__NaN__"
        elif math.isinf(val):
            return "__Infinity__" if val > 0 else "__-Infinity__"
        return val
    return val


def main():
    # Read specs from stdin
    input_data = json.loads(sys.stdin.read())
    specs = input_data["specs"]

    results = []

    for spec in specs:
        try:
            result = run_operation(spec)
            # Serialize to handle Infinity/NaN
            result = serialize_value(result)
            results.append(result)
        except Exception as e:
            print(f"Error running {spec['name']}: {e}", file=sys.stderr)
            results.append(None)

    # Output results as JSON
    print(json.dumps(results))


if __name__ == "__main__":
    main()
